### DISCUSSION TITLE
Incremental Backup Feature

### DISCUSSION THREAD

**Discussion by SleepyBearDad** [milvus-io/milvus#27634] [Q&A and General discussion]
Hi,

In the Zilliz 2.3 Beta release I read that incremental backup was added, is there an API reference or documentation for incremental backup?

Thanks,
Shahar

**Comment by yhmo** [comment]
Where did you see the "Zilliz 2.3 Beta release"?
We have a blog article that introduces the CDC tool(incremental backup), but it is about milvus, not the zilliz cloud. The CDC tool is not ready(almost to beta stage)
https://milvus.io/blog/unveiling-milvus-2-3-milestone-release-offering-support-for-gpu-arm64-cdc-and-other-features.md

**Comment by SleepyBearDad** [comment]
This is the blog post: https://zilliz.com/blog/milvus-2-3-beta-new-features-and-updates

We are using Milvus, not Zilliz cloud, but from the blog post I was under the impression that it applies to Milvus...

I have seen some discussions regarding the CDC tool, and have been looking at the source code. From how it looks, it seems that the CDC  is implemented as a sort of replication between 2 Milvus instances so it would require some tweaking to get it to do incremental backup, but entirely possible.

Is there someone working on this feature? I have seen this feature: https://github.com/milvus-io/milvus/issues/9685
But there hasn't been any activity lately.

There is also this: https://github.com/milvus-io/milvus/issues/16519 which mentions incremental backup based on rocksdb/pulsar which is like the CDC suggestion.

**Reply by yhmo** [reply]
The #9685 is about backup/restore, we already have this tool: https://github.com/zilliztech/milvus-backup.
The milvus-backup tool is static backup, it doesn't support increment backup. Currently, it is the official recommended tool to do backup/restore. The CDC tool is mainly for multi-cluster disaster recovery.

**Comment by SleepyBearDad** [comment]
Got it, so to implement an incremental backup using the current tools, I would need to implement a CDC-like tool that listens on kafka/pulsar + etcd, aggregate the changes from the full snapshot, and then use bulk API to restore the changes (+ sync with metadata changes such as create/delete collection).

A related question about the milvus API, are there thoughts to implement a quiesce API?
This will allow to perform backup from the k8s cluster by performing:
- quiesce
- flush
- PV backup (of the underlying minio volumes)

**Reply by yhmo** [reply]
The cdc tool watches the pulsar of the source cluster, and passes the requests to the target cluster. It also watches the etcd of the source cluster to get some information of collections/partitions. The CreateCollection/CreatePartition/DropCollection/DropPartition requests are passed to the target cluster by sdk(rpc call), the Insert/Delete requests are passed to the pulsar of the target cluster.

There is a feature in planning to "freeze" collection: https://github.com/milvus-io/milvus/issues/25363
A frozen collection cannot insert data, the semantics is similar to your quiesce API.

**Comment by SleepyBearDad** [comment]
Thanks for the answers!

Regarding the freeze API, it is not entirely equivalent to quiesce due to the constraint on the load API of loading an already loaded collection (https://milvus.io/docs/load_collection.md).

The goal of the quiesce is to allow applications to continue using the collection during the time it is frozen for read and for writes no error is received, only a delay in write completion (e.g. writes can be saved in kafka/pulsar but the data nodes will not start processing them).
An alternative design of quiesce, which will serve the purpose of allowing object store backup is to combine with flush and disable additional flush operations until the collection unquiesced. This will allow backing up the object store without concern for changes, this design has no impact on writes except when the flush is called for segments that exceed designated size.

Does this make sense as a feature?

**Reply by yhmo** [reply]
The complexity is more than expected. The internal compaction operation and index job also can change the object store even without flush actions. This api needs to stop all write actions to the object store.
Anyway, you can put your requirements here: https://github.com/milvus-io/milvus/discussions/6302

**Comment by SleepyBearDad** [comment]
Thanks, will post on the feature discussion!

A small additional question.
I have deployed milvus CDC, but it seems to attempt to watch the entire collection under etcd key <my release>/meta/root-coord/collection, but looking at the etcd keys, I only see keys for credential, database, fields and partitions.

Does this mean that the CDC can only watch for changes in the database/partition context?

**Reply by SleepyBearDad** [reply]
Just for clarity, the source (upstream) Milvus cluster has been installed using helm following: https://milvus.io/docs/install_cluster-milvusoperator.md

image is milvusdb/milvus:2.3

**Reply by SimFG** [reply]
The milvus cdc is developing. At present, CDC is still in an unavailable state, and some of its implementations may change at any time.

**Reply by SleepyBearDad** [reply]
Got it, thanks!

I will modify some code to make it work with version 2.3 but will also wait for the formal release of a more stable version.
